lib/dataset/sort.cpp

```diff
@@ -3,45 +3,83 @@
 /// @file
 /// @author Simon Heybrock
 #include "scipp/dataset/sort.h"
-#include "scipp/dataset/groupby.h"
+#include "scipp/core/tag_util.h"
+#include "scipp/dataset/extract.h"
 
 namespace scipp::dataset {
 
+namespace {
+
+constexpr auto nan_sensitive_less = [](const auto &a, const auto &b) {
+  if constexpr (std::is_floating_point_v<std::decay_t<decltype(a)>>)
+    if (std::isnan(b))
+      return !std::isnan(a);
+  return a < b;
+};
+
+template <class T> struct IndicesForSorting {
+  static Variable apply(const Variable &key, const SortOrder order) {
+    const auto size = key.dims()[key.dim()];
+    const auto values = key.values<T>();
+    std::vector<std::pair<T, scipp::index>> pairs;
+    pairs.reserve(size);
+    scipp::index i = 0;
+    for (const auto &value : key.values<T>())
+      pairs.emplace_back(value, i++);
```

Comment:
Maybe but this into a separate function. It leaks `i` and I had to check that it is not used anywhere below.